colossalai/nn/optimizer/adafactor.py

```diff
@@ -0,0 +1,228 @@
+import math
+import os 
+import torch
+from torch.optim import Optimizer
+
+
+# Adafactor 
+class Adafactor(Optimizer):
+    def __init__(
+        self,
+        params,
+        lr=None,
+        eps=(1e-30, 1e-3),
+        clip_threshold=1.0,
+        decay_rate=-0.8,
+        beta1=None,
+        weight_decay=0.0,
+        scale_parameter=True,
+        relative_step=True,
+        warmup_init=False,
+    ):
+        if lr is not None and relative_step:
+            raise ValueError("Cannot combine manual `lr` and `relative_step=True` options")
+        if warmup_init and not relative_step:
+            raise ValueError("`warmup_init=True` requires `relative_step=True`")
+
+        defaults = {
+            "lr": lr,
+            "eps": eps,
+            "clip_threshold": clip_threshold,
+            "decay_rate": decay_rate,
+            "beta1": beta1,
+            "weight_decay": weight_decay,
+            "scale_parameter": scale_parameter,
+            "relative_step": relative_step,
+            "warmup_init": warmup_init,
+        }
+        super().__init__(params, defaults)
+
+    @staticmethod
+    def _get_lr(param_group, param_state):
+        rel_step_sz = param_group["lr"]
+        if param_group["relative_step"]:
+            min_step = 1e-6 * param_state["step"] if param_group["warmup_init"] else 1e-2
+            rel_step_sz = min(min_step, 1.0 / math.sqrt(param_state["step"]))
+        param_scale = 1.0
+        if param_group["scale_parameter"]:
+            param_scale = max(param_group["eps"][1], param_state["RMS"])
+        return param_scale * rel_step_sz
```

Comment:
Is this proposed in the origin paper? It seems like a lr scheduler. I doubt whether we should implement it here

Reply:
It actually comes from the transformer.optim.adafactor (already add License), but is not involved in the calculations in the tests (default lr is set to None)

Reply:
Can you find which paper propose this?

Reply:
> Can you find which paper propose this?

For the "rel_step_sz ", it can be found in here:
![rel_step_sz](https://github.com/hpcaitech/ColossalAI/assets/54985467/4fe13dc1-c826-4bc8-95df-19384d5f8a42)
![rel_step_sz_2](https://github.com/hpcaitech/ColossalAI/assets/54985467/6ba032f0-e201-490a-be8f-a39e921df9d8)

But for the "param_scale"， I didn't find evidence of this in the original article